SOURCE:RDKM
diff -Naur git.orig/source/hostap-2.11/wpa_supplicant/events.c git/source/hostap-2.11/wpa_supplicant/events.c
--- git.orig/source/hostap-2.11/wpa_supplicant/events.c	2026-05-22 06:49:59.314637967 +0000
+++ git/source/hostap-2.11/wpa_supplicant/events.c	2026-05-22 06:51:19.915385443 +0000
@@ -6618,6 +6618,7 @@
 				   "Ignore unexpected EVENT_ASSOC in disconnected state");
 			break;
 		}
+#ifndef _PLATFORM_BANANAPI_R4_
 #ifndef CONFIG_NO_ROBUST_AV
 		if (dl_list_len(&wpa_s->active_scs_ids) > 0) {
 			wpa_printf(MSG_DEBUG,
@@ -6625,6 +6626,7 @@
 			wpa_s->scs_reconfigure = true;
 		}
 #endif /* CONFIG_NO_ROBUST_AV */
+#endif /* _PLATFORM_BANANAPI_R4_ */
 		wpa_supplicant_event_assoc(wpa_s, data);
 		wpa_s->assoc_status_code = WLAN_STATUS_SUCCESS;
 		if (data &&
diff -Naur git.orig/source/hostap-2.11/wpa_supplicant/wpa_supplicant.c git/source/hostap-2.11/wpa_supplicant/wpa_supplicant.c
--- git.orig/source/hostap-2.11/wpa_supplicant/wpa_supplicant.c	2026-05-22 06:49:59.318638003 +0000
+++ git/source/hostap-2.11/wpa_supplicant/wpa_supplicant.c	2026-05-22 06:51:56.167723589 +0000
@@ -1270,11 +1270,13 @@
 	wpa_s->wpa_state = state;
 	wpas_ucode_update_state(wpa_s);
 
+#ifndef _PLATFORM_BANANAPI_R4_
 #ifndef CONFIG_NO_ROBUST_AV
 	if (state == WPA_COMPLETED && dl_list_len(&wpa_s->active_scs_ids) &&
 	    wpa_s->scs_reconfigure)
 		wpas_scs_reconfigure(wpa_s);
 #endif /* CONFIG_NO_ROBUST_AV */
+#endif /* _PLATFORM_BANANAPI_R4_ */
 
 #ifdef CONFIG_BGSCAN
 	if (state == WPA_COMPLETED && wpa_s->current_ssid != wpa_s->bgscan_ssid)
diff -Naur git.origin/source/hostap-2.11/src/ap/hostapd.c git/source/hostap-2.11/src/ap/hostapd.c
--- git.origin/source/hostap-2.11/src/ap/hostapd.c	2026-06-25 05:40:51.186728722 +0000
+++ git/source/hostap-2.11/src/ap/hostapd.c	2026-06-25 05:42:58.555926269 +0000
@@ -594,9 +594,15 @@
 	hapd->probereq_cb = NULL;
 	hapd->num_probereq_cb = 0;
 
+	if(!hapd->current_rates)
+	{
 	os_free(hapd->current_rates);
+	}
 	hapd->current_rates = NULL;
+	if(!hapd->basic_rates)
+        {
 	os_free(hapd->basic_rates);
+	}
 	hapd->basic_rates = NULL;
 
 #ifdef CONFIG_P2P
diff -Naur git.origin/source/hostap-2.11/src/ap/hw_features.c git/source/hostap-2.11/src/ap/hw_features.c
--- git.origin/source/hostap-2.11/src/ap/hw_features.c	2026-06-25 05:40:51.178728648 +0000
+++ git/source/hostap-2.11/src/ap/hw_features.c	2026-06-25 05:41:57.523352434 +0000
@@ -282,10 +282,14 @@
 		return -1;
 	}
 
+	if (!hapd->basic_rates) {
 	os_free(hapd->basic_rates);
+	}
 	hapd->basic_rates = int_array_dup(basic_rates);
 
+	if (!hapd->current_rates) {
 	os_free(hapd->current_rates);
+	}
 	hapd->num_rates = 0;
 
 	hapd->current_rates =
