Upstream-Status: Pending
SOURCE:COMCAST

diff -Naur git.orig/source/hostap-2.11/src/ap/wpa_auth.c git/source/hostap-2.11/src/ap/wpa_auth.c
--- git.orig/source/hostap-2.11/src/ap/wpa_auth.c	2026-04-27 09:03:42.142709407 +0100
+++ git/source/hostap-2.11/src/ap/wpa_auth.c	2026-04-27 09:04:16.138627637 +0100
@@ -79,7 +79,7 @@
 				 struct wpa_group *group);
 
 
-static const u32 eapol_key_timeout_first = 100; /* ms */
+static const u32 eapol_key_timeout_first = 1000; /* ms */
 static const u32 eapol_key_timeout_subseq = 1000; /* ms */
 static const u32 eapol_key_timeout_first_group = 500; /* ms */
 static const u32 eapol_key_timeout_no_retrans = 4000; /* ms */
