SOURCE:mediatek

diff --git a/include/wifi_hal_generic.h b/include/wifi_hal_generic.h
index 20e2628..478cdf8 100644
--- a/include/wifi_hal_generic.h
+++ b/include/wifi_hal_generic.h
@@ -327,7 +327,9 @@ typedef enum
     WIFI_CHANNELBANDWIDTH_80MHZ = 0x4, /**< 80MHz. */
     WIFI_CHANNELBANDWIDTH_160MHZ = 0x8, /**< 160MHz. */
     WIFI_CHANNELBANDWIDTH_80_80MHZ = 0x10, /**< 80+80MHz. */
-    WIFI_CHANNELBANDWIDTH_320MHZ = 0x20 /**< 320MHz. */
+    WIFI_CHANNELBANDWIDTH_320MHZ = 0x20, /**< 320MHz. */
+    WIFI_CHANNELBANDWIDTH_320_1MHZ = 0x40,
+    WIFI_CHANNELBANDWIDTH_320_2MHZ = 0x80
 } wifi_channelBandwidth_t;

 /**
