Source: RDK
diff --git a/src/wifi_hal.c b/src/wifi_hal.c
index 4ae3daf..1592c31 100644
--- a/src/wifi_hal.c
+++ b/src/wifi_hal.c
@@ -492,6 +492,15 @@ INT wifi_hal_pre_init()
         wifi_hal_info_print("%s:%d: platfrom pre init\n", __func__, __LINE__);
         pre_init_fn();
     }
+
+#ifdef CONFIG_IEEE80211BE 
+void hostapd_wpa_event(void *ctx, enum wpa_event_type event,
+                       union wpa_event_data *data);
+
+     wpa_supplicant_event = hostapd_wpa_event;
+#endif
+
+
     return RETURN_OK;
 }
 
