Upstream-Status: Pending
diff --git a/src/wifi_hal_hostapd.c b/src/wifi_hal_hostapd.c
index 5ea8fc1..9e51245 100644
--- a/src/wifi_hal_hostapd.c
+++ b/src/wifi_hal_hostapd.c
@@ -1459,6 +1459,9 @@ int update_hostap_iface_flags(wifi_interface_info_t *interface)
 
 int update_hostap_iface(wifi_interface_info_t *interface)
 {
+    #if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+    struct hostapd_data *hdata = interface ? &interface->u.ap.hapd : NULL;
+#endif
     struct hostapd_iface   *iface;
     wifi_vap_info_t *vap;
     wifi_radio_info_t *radio;
@@ -1489,7 +1492,11 @@ int update_hostap_iface(wifi_interface_info_t *interface)
     struct eht_capabilities *drv_eht_cap;
 #endif // HOSTAPD_VERSION >= 211
 #endif // CONFIG_IEEE80211BE
-
+#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+    if (hdata == NULL) {
+        return RETURN_ERR;
+    }
+#endif // 
     if (interface == NULL) {
         return RETURN_ERR;
     }
@@ -1549,8 +1556,11 @@ int update_hostap_iface(wifi_interface_info_t *interface)
         }
         return RETURN_ERR;
     }
-
+#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+    hdata->basic_rates = radio->basic_rates[band];
+    #else
     iface->basic_rates = radio->basic_rates[band];
+#endif
 
     get_coutry_str_from_code(param->countryCode, country);
     iface->freq = ieee80211_chan_to_freq(country, param->operatingClass, param->channel);
@@ -1576,6 +1586,25 @@ int update_hostap_iface(wifi_interface_info_t *interface)
     mode = iface->current_mode;
 
     if ((strlen (vap->u.bss_info.preassoc.supported_data_transmit_rates) > 0) && strcmp(vap->u.bss_info.preassoc.supported_data_transmit_rates, "disabled")) {
+	    #if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+        if(hdata->current_cac_rates) {
+            os_free(hdata->current_cac_rates);
+        }
+        hdata->current_cac_rates = os_calloc(mode->num_rates, sizeof(struct hostapd_rate_data));
+        if (!hdata->current_cac_rates) {
+            wifi_hal_info_print("%s:%d Failed to allocate memory\n",__func__,__LINE__);
+            if(preassoc_supp_rates) {
+                os_free(preassoc_supp_rates);
+                preassoc_supp_rates = NULL;
+            }
+            if(preassoc_basic_rates) {
+                os_free(preassoc_basic_rates);
+                preassoc_basic_rates = NULL;
+            }
+            return RETURN_ERR;
+        }
+#else
+
         if(iface->current_cac_rates) {
             os_free(iface->current_cac_rates);
         }
@@ -1592,9 +1621,14 @@ int update_hostap_iface(wifi_interface_info_t *interface)
             }
             return RETURN_ERR;
         }
+#endif // BANANA_PI_PORT && KERNEL_6_12
     }
     else {
+	    #if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+        hdata->current_rates = radio->rate_data[band];
+#else
         iface->current_rates = radio->rate_data[band];
+#endif
     }
     wifi_hal_info_print("%s:%d: Interface: %s band: %d mode:%p (%d) has %d rates\n", __func__,
         __LINE__, interface->name, band, mode, mode->mode, mode->num_rates);
@@ -1623,7 +1657,11 @@ int update_hostap_iface(wifi_interface_info_t *interface)
 
     wifi_hal_info_print("%s:%d: Interface: %s band: %d mode:%p (%d) has %d rates\n", __func__,
         __LINE__, interface->name, band, mode, mode->mode, mode->num_rates);
+#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+    hdata->num_rates = 0;
+#else
     iface->num_rates = 0;
+#endif
     for (i = 0; i < mode->num_rates; i++) {
 /*
         if (iface->conf->supported_rates &&
@@ -1633,35 +1671,62 @@ int update_hostap_iface(wifi_interface_info_t *interface)
 */
 
         if (preassoc_supp_rates) {
+		#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+              if (!int_array_includes(preassoc_supp_rates,
+                      mode->rates[i])) {
+#else
               if (!hostapd_rate_found(preassoc_supp_rates,
                       mode->rates[i])) {
+#endif
                   continue;
               } else {
+		      #if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+                  rate = &hdata->current_cac_rates[hdata->num_rates];
+#else
                   rate = &iface->current_cac_rates[iface->num_rates];
+#endif
                   rate->rate = mode->rates[i];
               }
         } else {
+		#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+            rate = &hdata->current_rates[hdata->num_rates];
+#else
             rate = &iface->current_rates[iface->num_rates];
+#endif
             rate->rate = mode->rates[i];
         }
         if (preassoc_basic_rates) { 
+	#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+          if (int_array_includes(preassoc_basic_rates, rate->rate)) {
+#else
             if (hostapd_rate_found(preassoc_basic_rates, rate->rate)) {
+#endif
             rate->flags |= HOSTAPD_RATE_BASIC;
             }
             else {
               rate->flags &= ~(HOSTAPD_RATE_BASIC);
             }
         } else {
+    #if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+          if (int_array_includes(hdata->basic_rates, rate->rate)) { 
+#else
           if (hostapd_rate_found(iface->basic_rates, rate->rate)) {
+#endif
               rate->flags |= HOSTAPD_RATE_BASIC;
           }
           else {
             rate->flags &= ~(HOSTAPD_RATE_BASIC);
           }
         }
+	  #if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+        wifi_hal_dbg_print("%s:%d: RATE[%d] rate=%d flags=0x%x\n", __func__, __LINE__,
+            hdata->num_rates, rate->rate, rate->flags);
+        hdata->num_rates++;
+#else
         wifi_hal_dbg_print("%s:%d: RATE[%d] rate=%d flags=0x%x\n", __func__, __LINE__,
             iface->num_rates, rate->rate, rate->flags);
         iface->num_rates++;
+#endif
     }
     cf1 = iface->freq;
     freq1 = cf1;
diff --git a/src/wifi_hal_nl80211.c b/src/wifi_hal_nl80211.c
index 89383bf..3fd411b 100644
--- a/src/wifi_hal_nl80211.c
+++ b/src/wifi_hal_nl80211.c
@@ -5472,6 +5472,14 @@ static void wiphy_info_extended_capab(wifi_driver_data_t *drv,
 
         wifi_hal_dbg_print("%s:%d: nl80211: EML Capability: 0x%x MLD Capability: 0x%x\n", __func__,
             __LINE__, capa->eml_capa, capa->mld_capa_and_ops);
+#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+        if (tb1[NL80211_ATTR_EML_CAPABILITY] &&
+                    tb1[NL80211_ATTR_EXT_MLD_CAPA_AND_OPS])
+                        capa->ext_mld_capa_and_ops =
+                                nla_get_u16(tb1[NL80211_ATTR_EXT_MLD_CAPA_AND_OPS]);
+        wifi_hal_dbg_print("%s:%d: nl80211: Extended MLD Capabilities and Operations: 0x%x", __func__,
+            __LINE__, capa->ext_mld_capa_and_ops);
+#endif // BANANA_PI_PORT && KERNEL_6_12
 #endif /* CONFIG_IEEE80211BE */
 #endif /* HOSTAPD_VERSION >= 211 */
 
@@ -10075,6 +10083,15 @@ int nl80211_connect_sta(wifi_interface_info_t *interface)
         } else {
             interface->wpa_s.conf->sae_pwe = 1;
         }
+#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+        interface->wpa_s.current_ssid->pt = sae_derive_pt(interface->wpa_s.conf->sae_groups,
+            interface->wpa_s.current_ssid->ssid,
+            interface->wpa_s.current_ssid->ssid_len,
+            (const u8*) interface->wpa_s.current_ssid->sae_password,
+            os_strlen(interface->wpa_s.current_ssid->sae_password),
+            (const u8*) interface->wpa_s.current_ssid->sae_password_id,
+            interface->wpa_s.current_ssid->sae_password_id ? os_strlen(interface->wpa_s.current_ssid->sae_password_id) : 0);
+#else
 
         interface->wpa_s.current_ssid->pt = sae_derive_pt(interface->wpa_s.conf->sae_groups,
             interface->wpa_s.current_ssid->ssid,
@@ -10082,6 +10099,7 @@ int nl80211_connect_sta(wifi_interface_info_t *interface)
             interface->wpa_s.current_ssid->sae_password,
             os_strlen(interface->wpa_s.current_ssid->sae_password),
             interface->wpa_s.current_ssid->sae_password_id);
+#endif // BANANA_PI_PORT && KERNEL_6_12
     }
 
 #ifdef CONFIG_WIFI_EMULATOR
@@ -11873,8 +11891,13 @@ int wifi_drv_get_ext_capab(void *priv, enum wpa_driver_if_type type,
 
 #if HOSTAPD_VERSION >= 211
 #ifdef CONFIG_IEEE80211BE
+#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+static int wifi_drv_get_mld_capab(void *priv, enum wpa_driver_if_type type,
+                                 u16 *eml_capa, u16 *mld_capa_and_ops, u16 *ext_mld_capa_and_ops)
+#else
 static int wifi_drv_get_mld_capab(void *priv, enum wpa_driver_if_type type,
                                  u16 *eml_capa, u16 *mld_capa_and_ops)
+#endif // BANANA_PI_PORT && KERNEL_6_12
 {
     wifi_interface_info_t *interface;
     wifi_vap_info_t *vap;
@@ -11882,6 +11905,11 @@ static int wifi_drv_get_mld_capab(void *priv, enum wpa_driver_if_type type,
     wifi_driver_data_t *drv;
     enum nl80211_iftype nlmode;
     unsigned int i;
+#if defined(BANANA_PI_PORT) && defined(KERNEL_6_12)
+    if (!ext_mld_capa_and_ops) {
+        return -1;
+    }
+#endif // BANANA_PI_PORT && KERNEL_6_12
 
     if (!eml_capa || !mld_capa_and_ops) {
         return -1;
@@ -13381,13 +13409,13 @@ fail:
 }
 
 #ifdef BANANA_PI_PORT
-#if defined(KERNEL_6_6)
+#if defined(KERNEL_6_6) && !defined(KERNEL_6_12)
 int wifi_drv_set_wds_sta(void *priv, const u8 *addr, int aid, int val, const char *bridge_ifname,
     const char *ifname_wds, u32 radio_mask)
 #else
 int wifi_drv_set_wds_sta(void *priv, const u8 *addr, int aid, int val, const char *bridge_ifname,
     const char *ifname_wds)
-#endif //KERNEL_6_6
+#endif //KERNEL_6_6 && !KERNEL_6_12
 #else
 int wifi_drv_set_wds_sta(void *priv, const u8 *addr, int aid, int val, const char *bridge_ifname,
     char *ifname_wds)
@@ -15134,9 +15162,15 @@ int wifi_drv_if_remove(void *priv, enum wpa_driver_if_type type, const char *ifn
 }
 
 #ifdef BANANA_PI_PORT
+#if !defined(KERNEL_6_12)
+static int wifi_drv_if_add(void *priv, enum wpa_driver_if_type type, const char *ifname,
+     const u8 *addr, void *bss_ctx, void **drv_priv, char *force_ifname, u8 *if_addr,
+     const char *bridge, int use_existing, int setup_ap, int freq, u32 radio_mask)
+#else
 static int wifi_drv_if_add(void *priv, enum wpa_driver_if_type type, const char *ifname,
     const u8 *addr, void *bss_ctx, void **drv_priv, char *force_ifname, u8 *if_addr,
-    const char *bridge, int use_existing, int setup_ap, int freq, u32 radio_mask)
+    const char *bridge, int use_existing, int setup_ap, int freq)
+#endif // !KERNEL_6_12
 #else
 static int wifi_drv_if_add(void *priv, enum wpa_driver_if_type type, const char *ifname,
     const u8 *addr, void *bss_ctx, void **drv_priv, char *force_ifname, u8 *if_addr,
