Upstream-Status: Pending
Index: platform/rdkb_hal/src/platform/platform_hal.c
===================================================================
--- platform.orig/rdkb_hal/src/platform/platform_hal.c
+++ platform/rdkb_hal/src/platform/platform_hal.c
@@ -419,6 +419,8 @@ INT platform_hal_GetTotalMemorySize(ULONG *pulSize)
     char buf[64] = {0};
     char cmd[64] = {0};
     FILE *fp = NULL;
+    if (pulSize == NULL)
+        return RETURN_ERR;
 
     sprintf(cmd, "awk '/MemTotal/ {print $2}' /proc/meminfo > /tmp/total_Mem");    
     system(cmd);
