diff --git a/platform/banana-pi/platform.c b/platform/banana-pi/platform.c
index 3a960ee..00a41c1 100644
--- a/platform/banana-pi/platform.c
+++ b/platform/banana-pi/platform.c
@@ -40,6 +40,7 @@
 #define MAX_KEYPASSPHRASE_LEN 129
 #define MAX_SSID_LEN 33
 #define INVALID_KEY  "12345678"
+#define WL_MAXRATES_IN_SET		16
 
 int wifi_nvram_defaultRead(char *in,char *out);
 int _syscmd(char *cmd, char *retBuf, int retBufSize);
@@ -364,11 +365,101 @@ int wifi_setApRetrylimit(void *priv)
     return 0;
 }
 
-int platform_get_radio_caps(wifi_radio_index_t index)
+static void platform_get_radio_caps_2g(wifi_interface_info_t *interface)
 {
-    return 0;
 }
 
+static void platform_get_radio_caps_5g(wifi_interface_info_t *interface)
+{
+#if HOSTAPD_VERSION >= 211
+    static const u8 eht_phy_cap[EHT_PHY_CAPAB_LEN] = { 0x2c, 0x00, 0x1b, 0xe0, 0x00, 0xe7, 0x00,
+        0x7e, 0x00 };
+    static const u8 eht_mcs[] = { 0x44, 0x44, 0x44, 0x44, 0x44, 0x44 };
+    struct hostapd_iface *iface = &interface->u.ap.iface;
+
+for (int i = 0; i < iface->num_hw_features; i++) {
+        iface->hw_features[i].eht_capab[IEEE80211_MODE_AP].eht_supported = true;
+        iface->hw_features[i].eht_capab[IEEE80211_MODE_AP].mac_cap = 0x00c2;
+        memcpy(iface->hw_features[i].eht_capab[IEEE80211_MODE_AP].phy_cap, eht_phy_cap,
+            sizeof(eht_phy_cap));
+        memcpy(iface->hw_features[i].eht_capab[IEEE80211_MODE_AP].mcs, eht_mcs, sizeof(eht_mcs));
+#endif /* HOSTAPD_VERSION >= 211 */
+}
+
+}
+
+static void platform_get_radio_caps_6g(wifi_interface_info_t *interface)
+{
+#if HOSTAPD_VERSION >= 211
+    static const u8 eht_phy_cap[EHT_PHY_CAPAB_LEN] = { 0x2e, 0x00, 0x00, 0x60, 0x00, 0xe7, 0x00,
+        0x0e, 0x00 };
+    static const u8 eht_mcs[] = { 0x44, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44 };
+    struct hostapd_iface *iface = &interface->u.ap.iface;
+
+    for (int i = 0; i < iface->num_hw_features; i++) {
+        iface->hw_features[i].eht_capab[IEEE80211_MODE_AP].eht_supported = true;
+        iface->hw_features[i].eht_capab[IEEE80211_MODE_AP].mac_cap = 0x00c2;
+        memcpy(iface->hw_features[i].eht_capab[IEEE80211_MODE_AP].phy_cap, eht_phy_cap,
+            sizeof(eht_phy_cap));
+        memcpy(iface->hw_features[i].eht_capab[IEEE80211_MODE_AP].mcs, eht_mcs, sizeof(eht_mcs));
+#endif /* HOSTAPD_VERSION >= 211 */
+    }
+}
+
+int platform_get_radio_caps(wifi_radio_index_t index)
+{
+    wifi_radio_info_t *radio;
+    wifi_interface_info_t *interface;
+    radio = get_radio_by_rdk_index(index);
+    if (radio == NULL) {
+        wifi_hal_dbg_print("%s:%d failed to get radio for index\n", __func__, __LINE__);
+        return RETURN_ERR;
+    }
+
+#ifdef CONFIG_IEEE80211BE
+        radio->driver_data.capa.flags2 |= WPA_DRIVER_FLAGS2_MLO;
+
+        /*
+         * FIXME: Hardcodes eml_capa and mld_capa_and_ops because brcm does not provide
+         * this information and we have errors in hostap.
+         *
+         * Remove it when brcm makes the necessary changes.
+         */
+        radio->driver_data.iface_ext_capa[NL80211_IFTYPE_UNSPECIFIED].eml_capa =
+            (((8 << 11) & EHT_ML_EML_CAPA_TRANSITION_TIMEOUT_MASK) |
+                ((0 << 8) & EHT_ML_EML_CAPA_EMLMR_DELAY_MASK) |
+                ((0 << 7) & EHT_ML_EML_CAPA_EMLMR_SUPP) |
+                ((0 << 4) & EHT_ML_EML_CAPA_EMLSR_TRANS_DELAY_MASK) |
+                (((0 << 1) & EHT_ML_EML_CAPA_EMLSR_PADDING_DELAY_MASK) |
+                    ((1 << 0) & EHT_ML_EML_CAPA_EMLSR_SUPP)));
+
+        radio->driver_data.iface_ext_capa[NL80211_IFTYPE_UNSPECIFIED].mld_capa_and_ops =
+            ((0 << 12 & EHT_ML_MLD_CAPA_AAR_SUPP) |
+                ((15 << 7) & EHT_ML_MLD_CAPA_FREQ_SEP_FOR_STR_MASK) |
+                (EHT_ML_MLD_CAPA_TID_TO_LINK_MAP_ALL_TO_ALL |
+                    EHT_ML_MLD_CAPA_TID_TO_LINK_MAP_ALL_TO_ONE) |
+                ((0 << 4) & EHT_ML_MLD_CAPA_SRS_SUPP) |
+                ((MAX_NUM_MLD_LINKS - 1) & EHT_ML_MLD_CAPA_MAX_NUM_SIM_LINKS_MASK));
+#endif /* CONFIG_IEEE80211BE */
+
+    for (interface = hash_map_get_first(radio->interface_map); interface != NULL;
+        interface = hash_map_get_next(radio->interface_map, interface)) {
+
+        if (interface->vap_info.vap_mode == wifi_vap_mode_sta) {
+            continue;
+        }
+
+        if (strstr(interface->vap_info.vap_name, "private_ssid_2g")) {
+            platform_get_radio_caps_2g(interface);
+        } else if (strstr(interface->vap_info.vap_name, "private_ssid_5g")) {
+            platform_get_radio_caps_5g(interface);
+        } else if (strstr(interface->vap_info.vap_name, "private_ssid_6g")) {
+            platform_get_radio_caps_6g(interface);
+        }
+    }
+
+    return RETURN_OK;
+}
 
 INT wifi_sendActionFrameExt(INT apIndex, mac_address_t MacAddr, UINT frequency, UINT wait, UCHAR *frame, UINT len)
 {
