diff -Naur linux-6.6-origin/include/uapi/linux/cn_proc.h linux-6.6/include/uapi/linux/cn_proc.h
--- linux-6.6-origin/include/uapi/linux/cn_proc.h	2025-06-25 08:58:10.740217229 +0100
+++ linux-6.6/include/uapi/linux/cn_proc.h	2025-06-25 09:13:41.823330980 +0100
@@ -69,8 +69,7 @@
 
 static inline enum proc_cn_event valid_event(enum proc_cn_event ev_type)
 {
-	ev_type &= PROC_EVENT_ALL;
-	return ev_type;
+       return (enum proc_cn_event)(ev_type & PROC_EVENT_ALL);
 }
 
 /*
diff -Naur linux-6.6-origin/include/uapi/linux/connector.h linux-6.6/include/uapi/linux/connector.h
--- linux-6.6-origin/include/uapi/linux/connector.h	2025-06-25 08:58:10.744217260 +0100
+++ linux-6.6/include/uapi/linux/connector.h	2025-06-25 09:07:39.468560364 +0100
@@ -75,7 +75,7 @@
 
 	__u16 len;		/* Length of the following data */
 	__u16 flags;
-	__u8 data[];
+	__u8 data[0];
 };
 
 #endif /* _UAPI__CONNECTOR_H */
