diff --git a/source/scripts/init/service.d/lan_handler.sh b/source/scripts/init/service.d/lan_handler.sh
index e5ba1b33..7002165f 100755
--- a/source/scripts/init/service.d/lan_handler.sh
+++ b/source/scripts/init/service.d/lan_handler.sh
@@ -54,7 +54,7 @@ SERVICE_NAME="lan_handler"
 
 POSTD_START_FILE="/tmp/.postd_started"
 
-RPI_SPECIFIC=`grep -i "BOX_TYPE" /etc/device.properties  | cut -f2 -d=`
+RPI_SPECIFIC=`grep -nrw "BOX_TYPE" /etc/device.properties  | cut -f2 -d=`
 #args: router IP, subnet mask
 ap_addr() {
     if [ "$2" ]; then
