Source: COMCAST

diff -ruN git.orig/source/hostap-2.11/src/ap/dfs.c git/source/hostap-2.11/src/ap/dfs.c
--- git.orig/source/hostap-2.11/src/ap/dfs.c	2025-10-10 16:56:34.024160733 +0000
+++ git/source/hostap-2.11/src/ap/dfs.c	2025-10-10 17:03:41.625884044 +0000
@@ -155,7 +155,7 @@
 }
 
 
-static struct hostapd_channel_data *
+struct hostapd_channel_data *
 dfs_get_chan_data(struct hostapd_hw_modes *mode, int freq, int first_chan_idx)
 {
 	int i;
@@ -524,7 +524,7 @@
 }
 
 
-static struct hostapd_channel_data *
+struct hostapd_channel_data *
 dfs_get_valid_channel(struct hostapd_iface *iface,
 		      int *secondary_channel,
 		      u8 *oper_centr_freq_seg0_idx,
@@ -666,7 +666,7 @@
 }
 
 
-static int set_dfs_state_freq(struct hostapd_iface *iface, int freq, u32 state)
+int set_dfs_state_freq(struct hostapd_iface *iface, int freq, u32 state)
 {
 	struct hostapd_hw_modes *mode;
 	struct hostapd_channel_data *chan = NULL;
@@ -692,7 +692,7 @@
 }
 
 
-static int set_dfs_state(struct hostapd_iface *iface, int freq, int ht_enabled,
+int set_dfs_state(struct hostapd_iface *iface, int freq, int ht_enabled,
 			 int chan_offset, int chan_width, int cf1,
 			 int cf2, u32 state)
 {
@@ -758,7 +758,7 @@
 }
 
 
-static int dfs_are_channels_overlapped(struct hostapd_iface *iface, int freq,
+int dfs_are_channels_overlapped(struct hostapd_iface *iface, int freq,
 				       int chan_width, int cf1, int cf2)
 {
 	int start_chan_idx, start_chan_idx1;
@@ -1226,7 +1226,7 @@
 }
 
 
-static bool
+bool
 hostapd_dfs_is_background_event(struct hostapd_iface *iface, int freq)
 {
 	return (iface->drv_flags2 & WPA_DRIVER_FLAGS2_RADAR_BACKGROUND) &&
