##########################################
Date: Apr 16, 2024 1:00 PM
From:
Subject: Add RIFS mode parameter
Source: Comcast
Upstream-Status: Pending
Signed-off-by: Bogdan Bogush <bogdan.bogush@comcast.com>
##########################################

diff -ruN git.orig/source/hostap-2.11/src/ap/ap_config.h git/source/hostap-2.11/src/ap/ap_config.h
--- git.orig/source/hostap-2.11/src/ap/ap_config.h	2025-10-10 17:21:27.448207803 +0000
+++ git/source/hostap-2.11/src/ap/ap_config.h	2025-10-10 17:24:57.903086677 +0000
@@ -1184,6 +1184,7 @@
 	u8 vht_oper_centr_freq_seg0_idx;
 	u8 vht_oper_centr_freq_seg1_idx;
 	u8 ht40_plus_minus_allowed;
+	u8 ht_rifs;
 
 	/* Use driver-generated interface addresses when adding multiple BSSs */
 	u8 use_driver_iface_addr;
diff -ruN git.orig/source/hostap-2.11/src/ap/ieee802_11_ht.c git/source/hostap-2.11/src/ap/ieee802_11_ht.c
--- git.orig/source/hostap-2.11/src/ap/ieee802_11_ht.c	2025-10-10 17:21:27.440207845 +0000
+++ git/source/hostap-2.11/src/ap/ieee802_11_ht.c	2025-10-10 17:26:38.558550450 +0000
@@ -152,6 +152,10 @@
 
 	oper->primary_chan = hapd->iconf->channel;
 	oper->operation_mode = host_to_le16(hapd->iface->ht_op_mode);
+
+	if (hapd->iconf->ht_rifs == 1)
+               oper->ht_param |= HT_INFO_HT_PARAM_RIFS_MODE;
+
 	set_ht_param(hapd, oper);
 
 	vht_capabilities_info = host_to_le32(hapd->iface->current_mode->vht_capab);
