##########################################
Date: Apr 16, 2024 1:00 PM
From:
Subject: Allow power constraint tag for all bands
Source: Comcast
Upstream-Status: Pending
Signed-off-by: Bogdan Bogush <bogdan.bogush@comcast.com>
##########################################

diff -ruN git.orig/source/hostap-2.11/src/ap/beacon.c git/source/hostap-2.11/src/ap/beacon.c
--- git.orig/source/hostap-2.11/src/ap/beacon.c	2025-10-10 18:06:32.933768919 +0000
+++ git/source/hostap-2.11/src/ap/beacon.c	2025-10-10 18:10:09.396613065 +0000
@@ -131,13 +131,13 @@
 	u8 local_pwr_constraint = 0;
 	int dfs;
 
-	if (hapd->iface->current_mode == NULL ||
-	    hapd->iface->current_mode->mode != HOSTAPD_MODE_IEEE80211A)
+	if (hapd->iface->current_mode == NULL /*||
+            hapd->iface->current_mode->mode != HOSTAPD_MODE_IEEE80211A*/)	
 		return eid;
 
 	/* Let host drivers add this IE if DFS support is offloaded */
-	if (hapd->iface->drv_flags & WPA_DRIVER_FLAGS_DFS_OFFLOAD)
-		return eid;
+	/*if (hapd->iface->drv_flags & WPA_DRIVER_FLAGS_DFS_OFFLOAD)
+		return eid;*/
 
 	/*
 	 * There is no DFS support and power constraint was not directly
diff -ruN git.orig/source/hostap-2.11/src/ap/ieee802_11.c git/source/hostap-2.11/src/ap/ieee802_11.c
--- git.orig/source/hostap-2.11/src/ap/ieee802_11.c	2025-10-10 18:06:32.933768919 +0000
+++ git/source/hostap-2.11/src/ap/ieee802_11.c	2025-10-10 18:12:21.635906935 +0000
@@ -309,7 +309,7 @@
 	 * TODO: Also consider driver support for TPC to set Spectrum Mgmt bit
 	 */
 	if (hapd->iface->current_mode &&
-	    hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211A &&
+	  /*  hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211A &&*/
 	    (hapd->iconf->spectrum_mgmt_required || dfs))
 		capab |= WLAN_CAPABILITY_SPECTRUM_MGMT;
 
