##########################################
From: NarendraVarma_Dandu@comcast.com
Date: Jan 10, 2025 1:00 AM
Source: Comcast
##########################################

diff -ruN git.orig/source/hostap-2.11/src/ap/ieee802_11.c git/source/hostap-2.11/src/ap/ieee802_11.c
--- git.orig/source/hostap-2.11/src/ap/ieee802_11.c	2025-10-13 05:48:09.426991201 +0000
+++ git/source/hostap-2.11/src/ap/ieee802_11.c	2025-10-13 05:49:45.770472748 +0000
@@ -6168,6 +6168,11 @@
 		return;
 	}
 
+	os_free(sta->assoc_req);
+        sta->assoc_req = os_malloc(len);
+        os_memcpy(sta->assoc_req, (u8 *)mgmt, len);
+        sta->assoc_req_len = len;
+
 	if ((fc & WLAN_FC_RETRY) &&
 	    sta->last_seq_ctrl != WLAN_INVALID_MGMT_SEQ &&
 	    sta->last_seq_ctrl == seq_ctrl &&
diff -ruN git.orig/source/hostap-2.11/src/ap/sta_info.c git/source/hostap-2.11/src/ap/sta_info.c
--- git.orig/source/hostap-2.11/src/ap/sta_info.c	2025-10-13 05:48:09.426991201 +0000
+++ git/source/hostap-2.11/src/ap/sta_info.c	2025-10-13 05:52:36.585555652 +0000
@@ -532,6 +532,7 @@
 	if (sta->sae_pt)
 		sae_deinit_pt(sta->sae_pt);
 #endif
+	os_free(sta->assoc_req);
 
 	os_free(sta);
 }
diff -ruN git.orig/source/hostap-2.11/src/ap/sta_info.h git/source/hostap-2.11/src/ap/sta_info.h
--- git.orig/source/hostap-2.11/src/ap/sta_info.h	2025-10-13 05:48:09.418991244 +0000
+++ git/source/hostap-2.11/src/ap/sta_info.h	2025-10-13 05:53:22.113311215 +0000
@@ -339,6 +339,8 @@
 
 	u16 max_idle_period; /* if nonzero, the granted BSS max idle period in
 			      * units of 1000 TUs */
+	u8 *assoc_req;
+        size_t assoc_req_len;
 
 	u64 last_known_sta_id_timestamp;
 	bool mtk_sta;
