Source: RDK
--- git.origin/source/hostap-2.11/src/ap/beacon.c	2025-04-11 12:05:58.431681985 +0100
+++ git/source/hostap-2.11/src/ap/beacon.c	2025-04-11 12:08:12.817508589 +0100
@@ -3005,8 +3005,8 @@
 #endif /* CONFIG_IEEE80211BE */
 		params.freq = &freq;
 	}
-
-	params.freq->link_id = -1;
+        if (params.freq != NULL)
+		params.freq->link_id = -1;
 #ifdef CONFIG_IEEE80211BE
 	if (hapd->conf->mld_ap)
 		params.freq->link_id = hapd->mld_link_id;
